monai/data/utils.py

```diff
@@ -297,7 +296,24 @@ def list_data_collate(batch: Sequence):
         raise TypeError(re_str) from re
 
 
-def decollate_batch(batch, detach: bool = True):
+def _non_zipping_check(data):
+    """
+    utility function used by `decollate_batch`, to identify the largest batch size from the collated data.
+    returns batch_size and the list of non-iterable items.
+    """
+    batch_size, non_iterable = 0, []
+    for k, v in data.items() if isinstance(data, Mapping) else enumerate(data):
+        if not isinstance(v, Iterable) or isinstance(v, (str, bytes)) or (isinstance(v, torch.Tensor) and v.ndim == 0):
+            # Not running the usual list decollate here:
+            # don't decollate ['test', 'test'] into [['t', 't'], ['e', 'e'], ['s', 's'], ['t', 't']]
+            # torch.tensor(0) is iterable but iter(torch.tensor(0)) raises TypeError: iteration over a 0-d tensor
+            non_iterable.append(k)
+        elif hasattr(v, "__len__"):
```

Comment:
I think maybe we should only check the length of `Tensor` or `Numpy array`?
For example, below test code will output batch_size=`3` instead of `2`:
```py
data = {"image": torch.Tensor([1, 2]), "label": torch.Tensor([3, 4]), "meta_dict": {"filename": ["test1", "test2"], "affine": [0, 0], "original_shape": [1, 1]}}
print(_non_zipping_check(data)[0])
```

Thanks.

Reply:
that'll be wrong for `_non_zipping_check({"image": [1, 2, 3], "label": [1, 2]})`. this function is only used after `[decollate_batch(b, detach, pad=pad, fill_value=fill_value) for b in batch]`, I'll try to revise to improve the readability.

Reply:
Hi @wyli ,

Thanks for your quick update, I think the functionality is correct now.
But I feel it's still very confusing because `_non_zipping_check` calls `decollate_batch` and `decollate_batch` also calls `_non_zipping_check` in the recursive way.
I understand you like to make every APIs available to use directly, but In this case, I think `_non_zipping_check` should not be used independently, so maybe we don't need to set 2 separate functions, `_non_zipping_check` can be just an inline function of `decollate_batch` or we define smaller inline functions to make the logic easier to read?
What do you think?

Thanks.

Reply:
I'd like to keep it a separate function, and starting with underscore means a private function.

Reply:
It's OK to keep in 2 functions, is it possible to avoid `A calls B and B calls A`? Maybe put the `decollate` logic out of `_non_zipping_check`?

Thanks.

Reply:
That's my previous version, I think it causes confusions. This one also shows the fact that the batch size calculation is always for dict/list and after the decollate. Let me know if you have a better idea, I'm happy to change.

Reply:
OK, let's merge this PR to unblock v0.8 release first, I will also think about it, maybe we can improve it later.

Thanks.

Reply:
sure, this function is currently not very useful outside of this decollate process, perhaps we can improve it if we find some other useful cases.